ADC via DMA on STM32:  DMA device is null and `dma_config()` leads to Instruction Address Violation

Hello Developers!

_this is my first issue to Zephyr, or any other open source project! if i should put this somewhere else or I just do it wrong -- tell me please._

I have a problem with ADC over DMA on STM32F429i-DISC1 board running samples/drivers/adc sample. I get a hardfault on dma_config(), executing an undefined instruction (api->config()). I think that something is not fully initialized through the DT.

I had the same issue on a STM32G071 on my custom board on DMAMUX and I couldn't solve it, so I tried a SoC (F429) without the MUX, but the problem prevails. My workaround was to use HAL/LL for DMA temporarily, but I would like to switch back to Zephyr implementation whenever possible. Any suggestions are much appreciated. 

I include a debug screenshot, call stack and zephyr terminal output, together with the DTS overlay and prj.conf entries.

Specs: 
Board: STM32f429i-Disc1
Build: `west build -p -b stm32f429i_disc1 samples/drivers/adc`  
Debug and flash tool: JLink OCD
Development environment: VSCode + cortex-debug
ZephyrOS version: 3.5.0
SDK: 0.16.3_windows_x86_64

### prj.conf entries:
```
CONFIG_ADC=y
CONFIG_DMA=y
CONFIG_ADC_STM32_DMA=y
CONFIG_ADC_ASYNC=y
```

### Debug screenshot (VSCode + cortex-debug via JLink)

> **Note:** `api=dev->api;` while `dev` is 0x00! Therefore `api` struct contains nonsense, and calling a function from a pointer inside it will lead to provblems

![image](https://github.com/zephyrproject-rtos/zephyr/assets/71927208/68f477d6-7dd9-4033-a89a-6b2d27a11da3)


### Call Stack 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/71927208/4e47df54-e15b-44fe-9ecc-44beceb7c8ba)

### Terminal screenshot
![image](https://github.com/zephyrproject-rtos/zephyr/assets/71927208/3ebcaa26-7c35-4cb5-96c7-47b28af09880)


...and is my DTS overlay
```
 / {
    zephyr,user {
        io-channels = <&adc1 1>;//<&adc1 2 &adc1 1>;
    };
};

&adc1 {
    #address-cells = <1>;
    #size-cells = <0>;
    status = "okay";
    pinctrl-0 = <&gpio_a6_analog>;
    pinctrl-names = "default";
    st,adc-prescaler = <4>;
    st,adc-clock-source = <SYNC>;
    dmas = <&dma2 0 0 (STM32_DMA_PERIPH_TO_MEMORY | STM32_DMA_PERIPH_16BITS) 0x00>;
    dma-names="adcdma";
    channel@1 {
        reg = <1>;
        zephyr,gain = "ADC_GAIN_1";
        zephyr,reference = "ADC_REF_INTERNAL";
        zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
        zephyr,resolution = <12>;
    };

};

&pinctrl {
    gpio_a6_analog: gpio_a6_analog {
        pinmux = <STM32_PINMUX('A',6,ANALOG)>;
    };
    status = "okay";
};


&dma2 {
    status = "okay";
};
``` 

Hope putting this issue may help someone in the future

p.s. Used screenshots instead of text logs because I put the same problem on Discord forum before, and this is rather a copy of the post I made

A developer who has this same issue also mentioned that they the following warning. This would mean that `CONFIG_ADC_STM32_DMA=y` as expected, but somehow the dma `struct stream` is never filled.
```
z-workspace/zephyr/drivers/adc/adc_stm32.c:713:13: warning: 'dma_callback' defined but not used [-Wunused-function]
  713 | static void dma_callback(const struct device *dev, void *user_data,
      |             ^~~~~~~~~~~~
[142/142] Linking C executable zephyr/zephyr.elf
```
This will only happen if this `DT_INST_DMAS_HAS_NAME` check fails. (Which is an odd silent failure to have. It should be better)
https://github.com/zephyrproject-rtos/zephyr/blob/bc08e801c6b1416dcbbc679f5e25f801d438a531/drivers/adc/adc_stm32.c#L1553-L1556

Which is most likely caused by the driver currently requiring a _dmamux_ to function, and not _any_ dma.
https://github.com/zephyrproject-rtos/zephyr/blob/bc08e801c6b1416dcbbc679f5e25f801d438a531/drivers/adc/adc_stm32.c#L1604-L1609

Of the top of my head I'm not sure how this should be fixed.

--- a/drivers/adc/adc_stm32.c
+++ b/drivers/adc/adc_stm32.c
@@ -1521,39 +1521,39 @@ static void adc_stm32_irq_init(void)
 #define ADC_STM32_IRQ_CONFIG(index)
 #define ADC_STM32_IRQ_FUNC(index)					\
 	.irq_cfg_func = adc_stm32_irq_init,
-#define ADC_DMA_CHANNEL(id, dir, DIR, src, dest)
+#define ADC_DMA_CHANNEL(id, src, dest)
 
 #elif defined(CONFIG_ADC_STM32_DMA) /* !CONFIG_ADC_STM32_SHARED_IRQS */
 
-#define ADC_DMA_CHANNEL_INIT(index, name, dir_cap, src_dev, dest_dev)			\
+#define ADC_DMA_CHANNEL_INIT(index, src_dev, dest_dev)					\
 	.dma = {									\
-		.dma_dev = DEVICE_DT_GET(STM32_DMA_CTLR(index, name)),			\
-		.channel = DT_INST_DMAS_CELL_BY_NAME(index, name, channel),		\
+		.dma_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_IDX(index, 0)),		\
+		.channel = STM32_DMA_SLOT_BY_IDX(index, 0, channel),			\
 		.dma_cfg = {								\
-			.dma_slot = STM32_DMA_SLOT(index, name, slot),			\
+			.dma_slot = STM32_DMA_SLOT_BY_IDX(index, 0, slot),		\
 			.channel_direction = STM32_DMA_CONFIG_DIRECTION(		\
-				STM32_DMA_CHANNEL_CONFIG(index, name)),			\
+				STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),		\
 			.source_data_size = STM32_DMA_CONFIG_##src_dev##_DATA_SIZE(	\
-				STM32_DMA_CHANNEL_CONFIG(index, name)),			\
+				STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),		\
 			.dest_data_size = STM32_DMA_CONFIG_##dest_dev##_DATA_SIZE(	\
-				STM32_DMA_CHANNEL_CONFIG(index, name)),			\
+				STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),		\
 			.source_burst_length = 1,       /* SINGLE transfer */		\
 			.dest_burst_length = 1,         /* SINGLE transfer */		\
 			.channel_priority = STM32_DMA_CONFIG_PRIORITY(			\
-				STM32_DMA_CHANNEL_CONFIG(index, name)),			\
+				STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),		\
 			.dma_callback = dma_callback,					\
 			.block_count = 2,						\
 		},									\
 		.src_addr_increment = STM32_DMA_CONFIG_##src_dev##_ADDR_INC(		\
-			STM32_DMA_CHANNEL_CONFIG(index, name)),				\
+			STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),			\
 		.dst_addr_increment = STM32_DMA_CONFIG_##dest_dev##_ADDR_INC(		\
-			STM32_DMA_CHANNEL_CONFIG(index, name)),				\
+			STM32_DMA_CHANNEL_CONFIG_BY_IDX(index, 0)),			\
 	}
 
-#define ADC_DMA_CHANNEL(id, dir, DIR, src, dest)					\
-	COND_CODE_1(DT_INST_DMAS_HAS_NAME(id, dir),					\
-		    (ADC_DMA_CHANNEL_INIT(id, dir, DIR, src, dest)),			\
-		    (EMPTY))
+#define ADC_DMA_CHANNEL(id, src, dest)							\
+	COND_CODE_1(DT_INST_DMAS_HAS_IDX(id, 0),					\
+			(ADC_DMA_CHANNEL_INIT(id, src, dest)),				\
+			(/* Required for other adc instances without dma */))
 
 #define ADC_STM32_IRQ_CONFIG(index)					\
 static void adc_stm32_cfg_func_##index(void){ EMPTY }
@@ -1572,7 +1572,7 @@ static void adc_stm32_cfg_func_##index(void)				\
 }
 #define ADC_STM32_IRQ_FUNC(index)					\
 	.irq_cfg_func = adc_stm32_cfg_func_##index,
-#define ADC_DMA_CHANNEL(id, dir, DIR, src, dest)
+#define ADC_DMA_CHANNEL(id, src, dest)
 
 #endif /* CONFIG_ADC_STM32_DMA && CONFIG_ADC_STM32_SHARED_IRQS */
 
@@ -1605,7 +1605,7 @@ static struct adc_stm32_data adc_stm32_data_##index = {			\
 	ADC_CONTEXT_INIT_TIMER(adc_stm32_data_##index, ctx),		\
 	ADC_CONTEXT_INIT_LOCK(adc_stm32_data_##index, ctx),		\
 	ADC_CONTEXT_INIT_SYNC(adc_stm32_data_##index, ctx),		\
-	ADC_DMA_CHANNEL(index, dmamux, NULL, PERIPHERAL, MEMORY)	\
+	ADC_DMA_CHANNEL(index, PERIPHERAL, MEMORY)			\
 };									\
 									\
 PM_DEVICE_DT_INST_DEFINE(index, adc_stm32_pm_action);			\

--- a/drivers/adc/adc_stm32.c
+++ b/drivers/adc/adc_stm32.c
@@ -216,6 +216,39 @@ static bool init_irq = true;
 #endif
 
 #ifdef CONFIG_ADC_STM32_DMA
+static void adc_stm32_enable_dma_support(ADC_TypeDef *adc)
+{
+	/* Allow ADC to create DMA request and set to one-shot mode as implemented in HAL drivers */
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+
+#if defined(ADC_VER_V5_V90)
+	if (adc == ADC3) {
+		LL_ADC_REG_SetDMATransferMode(adc,
+			ADC3_CFGR_DMACONTREQ(LL_ADC_REG_DMA_TRANSFER_LIMITED));
+		LL_ADC_EnableDMAReq(adc);
+	} else {
+		LL_ADC_REG_SetDataTransferMode(adc,
+			ADC_CFGR_DMACONTREQ(LL_ADC_REG_DMA_TRANSFER_LIMITED));
+	}
+#elif defined(ADC_VER_V5_X)
+	LL_ADC_REG_SetDataTransferMode(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
+#else
+#error "Unsupported ADC version"
+#endif
+
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc) /* defined(CONFIG_SOC_SERIES_STM32H7X) */
+
+#error "The STM32F1 ADC + DMA is not yet supported"
+
+#else /* DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc) */
+
+	/* Default mechanism for other MCUs */
+	LL_ADC_REG_SetDMATransfer(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
+
+#endif
+}
+
 static int adc_stm32_dma_start(const struct device *dev,
 			       void *buffer, size_t channel_count)
 {
@@ -257,21 +290,7 @@ static int adc_stm32_dma_start(const struct device *dev,
 		return ret;
 	}
 
-	/* Allow ADC to create DMA request and set to one-shot mode,
-	 * as implemented in HAL drivers, if applicable.
-	 */
-#if defined(ADC_VER_V5_V90)
-	if (adc == ADC3) {
-		LL_ADC_REG_SetDMATransferMode(adc,
-			ADC3_CFGR_DMACONTREQ(LL_ADC_REG_DMA_TRANSFER_LIMITED));
-		LL_ADC_EnableDMAReq(adc);
-	} else {
-		LL_ADC_REG_SetDataTransferMode(adc,
-			ADC_CFGR_DMACONTREQ(LL_ADC_REG_DMA_TRANSFER_LIMITED));
-	}
-#elif defined(ADC_VER_V5_X)
-	LL_ADC_REG_SetDataTransferMode(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
-#endif
+	adc_stm32_enable_dma_support(adc);
 
 	data->dma_error = 0;
 	ret = dma_start(data->dma.dma_dev, data->dma.channel);

--- a/include/zephyr/drivers/dma/dma_stm32.h
+++ b/include/zephyr/drivers/dma/dma_stm32.h
@@ -33,8 +33,10 @@
 /* macro for dma slot (only for dma-v1 or dma-v2 types) */
 #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_v2bis)
 #define STM32_DMA_SLOT(id, dir, slot) 0
+#define STM32_DMA_SLOT_BY_IDX(id, idx, slot)
 #else
 #define STM32_DMA_SLOT(id, dir, slot) DT_INST_DMAS_CELL_BY_NAME(id, dir, slot)
+#define STM32_DMA_SLOT_BY_IDX(id, idx, slot) DT_INST_DMAS_CELL_BY_IDX(id, idx, slot)
 #endif
 
 #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_v2) || \
@@ -50,6 +52,8 @@
 		DT_INST_DMAS_CTLR_BY_NAME(id, dir)
 #define STM32_DMA_CHANNEL_CONFIG(id, dir)					\
 		DT_INST_DMAS_CELL_BY_NAME(id, dir, channel_config)
+#define STM32_DMA_CHANNEL_CONFIG_BY_IDX(id, idx)				\
+		DT_INST_DMAS_CELL_BY_IDX(id, idx, channel_config)
 
 /* macros for channel-config */
 /* direction defined on bits 6-7 */

--- a/tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf
+++ b/tests/drivers/adc/adc_dma/boards/nucleo_l476rg.conf
@@ -1,0 +1,8 @@
+#
+# Copyright (c) 2023 Hein Wessels, Nobleo Technology
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_ADC_STM32_DMA=y
+CONFIG_ADC_ASYNC=y

--- a/tests/drivers/adc/adc_dma/boards/nucleo_l476rg.overlay
+++ b/tests/drivers/adc/adc_dma/boards/nucleo_l476rg.overlay
@@ -1,0 +1,17 @@
+/*
+ * Copyright (c) 2023 Hein Wessels, Nobleo Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&adc1 {
+	dmas = < &dma1 1 0 (STM32_DMA_PERIPH_RX |  STM32_DMA_MEM_16BITS | STM32_DMA_PERIPH_16BITS) >;
+	dma-names = "dma";
+
+	#address-cells = <1>;
+	#size-cells = <0>;
+};
+
+test_dma: &dma1 {
+    status = "okay";
+};

--- a/tests/drivers/adc/adc_dma/src/test_adc.c
+++ b/tests/drivers/adc/adc_dma/src/test_adc.c
@@ -61,6 +61,17 @@
 #define ADC_2ND_CHANNEL_ID 7
 #define ALIGNMENT 32
 
+#elif defined(CONFIG_BOARD_NUCLEO_L476RG)
+
+#define ADC_DEVICE_NODE DT_INST(0, st_stm32_adc)
+#define ADC_RESOLUTION 12
+#define ADC_GAIN ADC_GAIN_1
+#define ADC_REFERENCE ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME ADC_ACQ_TIME_DEFAULT
+#define ADC_1ST_CHANNEL_ID 1
+#define ADC_2ND_CHANNEL_ID 7
+#define ALIGNMENT 32
+
 #endif
 
 /* Invalid value that is not supposed to be written by the driver. It is used

